### PR TITLE
add option `smart-group-indent`

### PR DESCRIPTION
**Pull request by scalder27**
This option purpose to prevent adding empty lines between single rule groups.
It`s just look awful when a small neat three lines css-rule becomes the five lines one.

`{'sort-order': [ ['display'], ['position'], ['font-size'] ], 'smart-group-indent': true}`

``` css
/* before */
div {
    display: inline-block;

    position: absolute;

    font-size: 1.2em;
}

/* after */
div {
    display: inline-block;
    position: absolute;
    font-size: 1.2em;
}
```

### CODE REVIEW

**Comment by scalder27** [comment]
And, for the sake of clarity:

`{'sort-order': [ ['display'], ['position', 'top'], ['font-size'] ], 'smart-group-indent': true}`


```css
/* before */
div {
    display: inline-block;
    position: absolute;
    top: 5px;
    font-size: 1.2em;
}

/* after */
div {
    display: inline-block;

    position: absolute;
    top: 5px;

    font-size: 1.2em;
}
```

**Comment by scalder27** [comment]
@hudochenkov, what are you think about such an option? The option name is a bit sloppy, but nonetheless... =)

File: index.js [hunk 1]
```diff
@@ -161,6 +161,7 @@ function fetchAllCommentsAfterNode(comments, nextNode, node) {
 
 module.exports = postcss.plugin('postcss-sorting', function (opts) {
     var linesBetweenChildrenRules = getLinesBetweenChildrenFromOptions(opts);
+    var smartGroupIndentation = !!(opts || {})['smart-group-indent'] || false;
```
Comment (hudochenkov) [inline index.js]: Not explicit type conversion. Should be more straightforward and clear. Complex one-liners are not good for maintainability.

File: index.js [hunk 2]
```diff
@@ -226,7 +227,16 @@ module.exports = postcss.plugin('postcss-sorting', function (opts) {
                     var prevNode = node.prev();
 
                     if (prevNode && node.raws.before) {
-                        if (node.groupIndex > prevNode.groupIndex) {
+                        var nextNode = node.next();
+                        var preventLineBreaking = smartGroupIndentation &&
+                            node.type !== 'rule' &&
+                            prevNode.type !== 'rule' &&
+                            node.groupIndex > prevNode.groupIndex &&
```
Comment (hudochenkov) [inline index.js]: Not needed because this is checking in the following condition.

**Comment by hudochenkov** [comment]
I've modified your test CSS and got an unexpected result.

```css
/* Before: */

div {
    position: absolute;
    top: 10px;
    display: inline-block;
    font-size: 1.2em;
}

/* After: */

div {
    display: inline-block;
    position: absolute;
    top: 10px;

    font-size: 1.2em;
}

/* Expected: */

div {
    display: inline-block;
    position: absolute;
    top: 10px;
    font-size: 1.2em;
}
```

Based on the example above we see that this logic isn't entirely correct. `preventLineBreaking` is a tricky condition :) I think the logic should be like this: if `smartGroupIndentation` is enabled, and both `node` and `prevNode` are `decl`, and both `node` and `prevNode` are single nodes in their groups, then empty lines is not added to `node.raws.before`.

**Comment by hudochenkov** [comment]
I think option name should be `groupSingleDeclarations`. What do you think?

**Comment by scalder27** [comment]
Hello! Sorry, that I`ve not answered yet. Your option name is a way better!
I`ll do all fixes you`ve mentioned, and there must will be one more. About the next node is being a comment.

**Comment by hudochenkov** [comment]
@scalder27 it doesn't matter what the next node is because we check only current and previous :)

**Comment by scalder27** [comment]
@hudochenkov current and previous, it doesn`t exactly what I'm trying to achieve. This way it merge single declaration group with a multi declarations one.


```
{
    'sort-order': [ ['display', 'padding'], ['position'], ['color', 'font-size'] ],
    'group-single-declarations': true
}
```

```css
/* instead of */
div {
    display: inline-block;
    padding: 20px 10px;

    position: absolute;

    color: red;
    font-size: 1.2em;
}

/* it will be */
div {
    display: inline-block;
    padding: 20px 10px;

    position: absolute;
    color: red;
    font-size: 1.2em;
}
```

**Comment by hudochenkov** [comment]
Why it should be grouped with the next group and not with previous? Empty line in the whole plugin mean something like “start of the new group whatever this group is, except the first one in the rule”.

**Comment by hudochenkov** [comment]
What if user really wants to separate one property like this `position`:

```js
{
    'sort-order': [ ['display', 'padding'], ['position'], ['color', 'font-size'] ],
    'group-single-declarations': true
}
```

I think this is good output with the option enabled:

```css
div {
    display: inline-block;
    padding: 20px 10px;

    position: absolute;

    color: red;
    font-size: 1.2em;
}
```

And this is good, because first and third groups contain two nodes:

```css
div {
    /* comment */
    padding: 20px 10px;

    position: absolute;

    /* comment */
    font-size: 1.2em;
}
```

But this is not good, two following groups with one node (`decl`) in them:

```css
div {
    padding: 20px 10px;

    position: absolute;

    color: red;
    font-size: 1.2em;
}
```

And should be:

```css
div {
    padding: 20px 10px;
    position: absolute;

    color: red;
    font-size: 1.2em;
}
```

Rather (this is what you suggest as I understand):

```css
div {
    padding: 20px 10px;
    position: absolute;
    color: red;
    font-size: 1.2em;
}
```

So I still prefer to group only single declarations together, rather glued these single declarations to group with 2+ declarations in it. Otherwise, it would be very unpredictable output with the option enabled. It still would unpredictable though, but not so much.

**Comment by scalder27** [comment]
Indeed, we're talking about the same expectations!
My example above (`it will be` part) was just an illustration about what we'll get (and such a result is defenitely wrong) in a case when we are not aware about the next node. The previus node is single, is the current one single? And there can be a tricky case when the next node is an inline comment.

```
div {
    display: inline-block;
    padding: 20px 10px;

    position: absolute; /* comment */

    font-size: 1.2em;
}
```

And, currently, there are more other caveats:

1) the previous node isn't `decl`
```
div {   
    padding: 20px 10px;  /* comment */

    position: absolute;
    font-size: 1.2em;
}
```

2) we doesn't now exactly whether the previous node is a single. A `propertyIndex` for the previous node now equals 0.
```
div {
    /* comment */
    padding: 20px 10px;

    position: absolute;
    font-size: 1.2em;
}
```

All of these stuffs just should be taken into account. And resolved =)

**Comment by hudochenkov** [comment]
Great point about the inline comment!

**Pull request closed without merging**